swap_meet/vendor.py

```diff
@@ -1,2 +1,233 @@
+
+    swap_by_newest(other):
+        swaps your's and other vendor's newest items
+        returns true if swap is successful
+    """
+
+
+    def __init__(self, inventory = None): #do not set keyword argument to list bc it is a mutable type
+        """
+        Constructs all the necessary attributes for the vendor object.
+
+        Parameters
+        ----------
+            inventory : list
+                list of strings representing items in inventory
+        """
+
+        if not inventory:
+            self.inventory = []
+        else:
+            self.inventory = inventory
+
+
+    def add(self, item):
+        """
+        Appends the item to the inventory list.
+
+        Parameters
+        ----------
+        item : object
+
+        Returns
+        -------
+        item
+        """
+        self.inventory.append(item)
+        return item
+
+    
+    def remove(self, item):
+        """
+        Removes item from inventory list.
+
+        Parameters
+        ----------
+        item : str
+
+        Returns
+        -------
+        item or None
+        """
+        try:
+            self.inventory.remove(item)
+            return item
+        except ValueError:
+            return False
+    
+
+    def get_by_category(self, category):
+        """
+        Creates a list of items from inventory whose category attributes match the input category. 
+
+        Parameters
+        ----------
+        category : str
+
+        Returns
+        -------
+        items: list
+        """
+
+        return [item for item in self.inventory if item.category == category]
+    
+
+    def swap_items(self, vendor, my_item, vendor_item):
+        """
+        Removes and appends 2nd and 3rd parameters from my inventory respectively. Appends and removes 2nd and 3rd parameters from vendor's inventory respectively.
+
+        Parameters
+        -------
+        vendor: composite object
+        my item: component object 
+        vendor_item: component object
+
+        Returns
+        -------
+        self.inventory: object
+        
+        """
+        if my_item in self.inventory and vendor_item in vendor.inventory:
+                self.remove(my_item) 
+                self.add(vendor_item)            
+                vendor.remove(vendor_item)       
+                vendor.add(my_item) 
+
+                return True
+
+
+    def swap_first_item(self, vendor):
+        """
+        Removes and appends first item from vendor and first item vendor_to_swap_with from my inventory respectively.Removes and appends first item from vendor_to_swap_with and first item vendor from vendor's inventory respectively.
+
+        Parameters
+        -------
+        vendor: object
+
+        Returns
+        -------
+        boolean
+        """
+        try:
+            return self.swap_items(vendor, self.inventory[0], vendor.inventory[0])
+        except IndexError:
+            return False
```

Comment:
👍  Great use of helper functions and use of try/except!

Reply:
Currently, swapping any items has a time complexity of O(n) due to the `remove` method in `swap_items`. We can make a subtle improvement to O(1) by using multiple-assignment instead:

```
        self.inventory[0], vendor.inventory[0] = vendor.inventory[0], self.inventory[0]
```        

Here's more info on Python multiple-assignment: https://www.w3schools.com/python/gloss_python_assign_value_to_multiple_variables.asp